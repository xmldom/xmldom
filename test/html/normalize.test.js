'use strict';

const { getTestParser } = require('../get-test-parser');

describe('html normalizer', () => {
	it.each([
		'<div>&amp;&lt;123&456<789;&&</div>',
		'<div><123e>&<a<br/></div>',
		'<div>&nbsp;&copy;&nbsp&copy</div>',
		'<html xmlns:x="1"><body/></html>',
		'<html test="a<b && a>b && \'&amp;&&\'"/>',
		'<div test="alert(\'<br/>\')"/>',
		'<div test="a<b&&a< c && a>d"></div>',
		'<div a=& bb c d=123&&456/>',
		'<div a=& a="&\'\'" b/>',
		'<html test="123"/>',
		'<r><Label onClick="doClick..>Hello, World</Label></r>',
		'<Label onClick=doClick..">Hello, World</Label>',
	])('%s', (xml) => {
		const { errors, parser } = getTestParser();

		const actual = parser.parseFromString(xml, 'text/html').toString();

		expect({ actual, ...errors }).toMatchSnapshot();
	});

	it.each([
		'<html><meta><link><img><br><hr><input></html>',
		'<html title =1/2></html>',
		'<html title= 1/>',
		'<html title = 1/>',
		'<html title/>',
	])('unclosed html %s', (xml) => {
		const { errors, parser } = getTestParser();

		const actual = parser.parseFromString(xml, 'text/html').toString();

		expect({ actual, ...errors }).toMatchSnapshot();
	});

	Array.from(['text/xml', 'text/html']).forEach((mimeType) => {
		it.each([
			'<script>alert(a<b&&c?"<br>":">>");</script>',
			'<script>alert(a<b&&c?"<br/>":">>");</script>',
			'<script src="./test.js"/>',
			'<textarea>alert(a<b&&c?"<br>":">>");</textarea>',
			'<input type="button" disabled></input>',
			'<input type="checkbox" checked></input>',
			'<option selected></option>',
			'<ul><li>abc<li>def</ul>',
		])(`${mimeType}: script %s`, (xml) => {
			const { errors, parser } = getTestParser();

			const actual = parser.parseFromString(xml, mimeType).toString();

			expect({ actual, ...errors }).toMatchSnapshot();
		});
	});

	it.each([
		`<html xmlns="http://www.w3.org/1999/xhtml"><script>let message = " &amp; ETH";</script></html>`,
		`<html><script>let message = " &amp; ETH";</script></html>`,
	])(`should map entity in %s`, (xml) => {
		const { parser } = getTestParser();

		const actual = parser.parseFromString(xml, 'application/xml');

		expect(actual.documentElement.firstChild.textContent).toBe('let message = " & ETH";');
	});
	it.each([
		`<html xmlns="http://www.w3.org/1999/xhtml"><script>let message = " &amp; ETH";</script></html>`,
		`<html><script>let message = " &amp; ETH";</script></html>`,
	])(`should not map entity in %s`, (xml) => {
		const { parser } = getTestParser();

		const actual = parser.parseFromString(xml, 'text/html');

		expect(actual.documentElement.firstChild.textContent).toBe('let message = " &amp; ETH";');
	});

	it('European entities', () => {
		const { errors, parser } = getTestParser();

		const actual = parser
			.parseFromString(
				'<div>&Auml;&auml;&Aring;&aring;&AElig;&aelig;&Ouml;&ouml;&Oslash;&oslash;&szlig;&Uuml;&uuml;&euro;</div>',
				'text/html'
			)
			.toString();

		expect({ actual, ...errors }).toMatchObject({
			// For the future, it may be nicer to use \uxxxx in the assert strings
			// rather than pasting in multi-byte UTF-8 Unicode characters
			actual: '<div xmlns="http://www.w3.org/1999/xhtml">ÄäÅåÆæÖöØøßÜü€</div>',
		});
	});
	it('European entities xml', () => {
		const { errors, parser } = getTestParser();

		const actual = parser
			.parseFromString(
				'<div>&Auml;&auml;&Aring;&aring;&AElig;&aelig;&Ouml;&ouml;&Oslash;&oslash;&szlig;&Uuml;&uuml;&euro;</div>',
				'text/xml'
			)
			.toString();

		expect({ actual, ...errors }).toMatchObject({
			// For the future, it may be nicer to use \uxxxx in the assert strings
			// rather than pasting in multi-byte UTF-8 Unicode characters
			actual:
				'<div>&amp;Auml;&amp;auml;&amp;Aring;&amp;aring;&amp;AElig;&amp;aelig;&amp;Ouml;&amp;ouml;&amp;Oslash;&amp;oslash;&amp;szlig;&amp;Uuml;&amp;uuml;&amp;euro;</div>',
		});
	});
});
