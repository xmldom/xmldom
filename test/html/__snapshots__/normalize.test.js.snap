// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html normalizer <Label onClick=doClick..">Hello, World</Label> 1`] = `
Object {
  "actual": "<Label onClick=\\"doClick..\\" xmlns=\\"http://www.w3.org/1999/xhtml\\">Hello, World</Label>",
  "warning": Array [
    "[xmldom warning]	attribute \\"onClick\\" missed start quot(\\")!!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer <div a=& a="&''" b/> 1`] = `
Object {
  "actual": "<div a=\\"&amp;''\\" b=\\"b\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></div>",
  "fatalError": Array [
    "[xmldom fatalError]	Attribute a redefined
@#[line:1,col:1]",
  ],
  "warning": Array [
    "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!
@#[line:1,col:1]",
    "[xmldom warning]	attribute \\"b\\" missed value!! \\"b\\" instead!!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer <div a=& bb c d=123&&456/> 1`] = `
Object {
  "actual": "<div a=\\"&amp;\\" bb=\\"bb\\" c=\\"c\\" d=\\"123&amp;&amp;456\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></div>",
  "warning": Array [
    "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!
@#[line:1,col:1]",
    "[xmldom warning]	attribute \\"bb\\" missed value!! \\"bb\\" instead2!!
@#[line:1,col:1]",
    "[xmldom warning]	attribute \\"c\\" missed value!! \\"c\\" instead2!!
@#[line:1,col:1]",
    "[xmldom warning]	attribute \\"123&&456\\" missed quot(\\")!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer <div test="a<b&&a< c && a>d"></div> 1`] = `
Object {
  "actual": "<div test=\\"a&lt;b&amp;&amp;a&lt; c &amp;&amp; a>d\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></div>",
}
`;

exports[`html normalizer <div test="alert('<br/>')"/> 1`] = `
Object {
  "actual": "<div test=\\"alert('&lt;br/>')\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></div>",
}
`;

exports[`html normalizer <div>&amp;&lt;123&456<789;&&</div> 1`] = `
Object {
  "actual": "<div xmlns=\\"http://www.w3.org/1999/xhtml\\">&amp;&lt;123&amp;456&lt;789;&amp;&amp;</div>",
  "error": Array [
    "[xmldom error]	element parse error: Error: invalid tagName:789;&&<
@#[line:1,col:22]",
  ],
}
`;

exports[`html normalizer <div>&nbsp;&copy;&nbsp&copy</div> 1`] = `
Object {
  "actual": "<div xmlns=\\"http://www.w3.org/1999/xhtml\\"> ©&amp;nbsp&amp;copy</div>",
}
`;

exports[`html normalizer <div><123e>&<a<br/></div> 1`] = `
Object {
  "actual": "<div xmlns=\\"http://www.w3.org/1999/xhtml\\">&lt;123e>&amp;&lt;a<br/></div>",
  "error": Array [
    "[xmldom error]	element parse error: Error: invalid tagName:123e
@#[line:1,col:6]",
    "[xmldom error]	element parse error: Error: invalid tagName:a<br
@#[line:1,col:13]",
  ],
}
`;

exports[`html normalizer <html test="123"/> 1`] = `
Object {
  "actual": "<html test=\\"123\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
}
`;

exports[`html normalizer <html test="a<b && a>b && '&amp;&&'"/> 1`] = `
Object {
  "actual": "<html test=\\"a&lt;b &amp;&amp; a>b &amp;&amp; '&amp;&amp;&amp;'\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
}
`;

exports[`html normalizer <html xmlns:x="1"><body/></html> 1`] = `
Object {
  "actual": "<html xmlns:x=\\"1\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"><body></body></html>",
}
`;

exports[`html normalizer <r><Label onClick="doClick..>Hello, World</Label></r> 1`] = `
Object {
  "actual": "<r xmlns=\\"http://www.w3.org/1999/xhtml\\">&lt;Label onClick=\\"doClick..>Hello, World</r>",
  "error": Array [
    "[xmldom error]	element parse error: Error: attribute value no end '\\"' match
@#[line:1,col:4]",
  ],
}
`;

exports[`html normalizer text/html: script <script src="./test.js"/> 1`] = `
Object {
  "actual": "<script src=\\"./test.js\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></script>",
}
`;

exports[`html normalizer text/html: script <script>alert(a<b&&c?"<br/>":">>");</script> 1`] = `
Object {
  "actual": "<script xmlns=\\"http://www.w3.org/1999/xhtml\\">alert(a<b&&c?\\"<br/>\\":\\">>\\");</script>",
}
`;

exports[`html normalizer text/html: script <script>alert(a<b&&c?"<br>":">>");</script> 1`] = `
Object {
  "actual": "<script xmlns=\\"http://www.w3.org/1999/xhtml\\">alert(a<b&&c?\\"<br>\\":\\">>\\");</script>",
}
`;

exports[`html normalizer text/html: script <textarea>alert(a<b&&c?"<br>":">>");</textarea> 1`] = `
Object {
  "actual": "<textarea xmlns=\\"http://www.w3.org/1999/xhtml\\">alert(a&lt;b&amp;&amp;c?\\"&lt;br>\\":\\">>\\");</textarea>",
}
`;

exports[`html normalizer text/xml: script <script src="./test.js"/> 1`] = `
Object {
  "actual": "<script src=\\"./test.js\\"/>",
}
`;

exports[`html normalizer text/xml: script <script>alert(a<b&&c?"<br/>":">>");</script> 1`] = `
Object {
  "actual": "<script>alert(a&lt;b&amp;&amp;c?\\"<br/>\\":\\">>\\");</script>",
  "error": Array [
    "[xmldom error]	element parse error: Error: attribute value must after \\"=\\"
@#[line:1,col:16]",
  ],
}
`;

exports[`html normalizer text/xml: script <script>alert(a<b&&c?"<br>":">>");</script> 1`] = `
Object {
  "actual": "<script>alert(a&lt;b&amp;&amp;c?\\"<br/>\\":\\">>\\");</script>",
  "error": Array [
    "[xmldom error]	element parse error: Error: attribute value must after \\"=\\"
@#[line:1,col:16]",
  ],
  "warning": Array [
    "[xmldom warning]	unclosed xml attribute
@#[line:1,col:23]",
  ],
}
`;

exports[`html normalizer text/xml: script <textarea>alert(a<b&&c?"<br>":">>");</textarea> 1`] = `
Object {
  "actual": "<textarea>alert(a&lt;b&amp;&amp;c?\\"<br/>\\":\\">>\\");</textarea>",
  "error": Array [
    "[xmldom error]	element parse error: Error: attribute value must after \\"=\\"
@#[line:1,col:18]",
  ],
  "warning": Array [
    "[xmldom warning]	unclosed xml attribute
@#[line:1,col:25]",
  ],
}
`;

exports[`html normalizer unclosed html <html title = 1/> 1`] = `
Object {
  "actual": "<html title=\\"1\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
  "warning": Array [
    "[xmldom warning]	attribute \\"1\\" missed quot(\\")!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer unclosed html <html title =1/2></html> 1`] = `
Object {
  "actual": "<html title=\\"1/2\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
  "warning": Array [
    "[xmldom warning]	attribute \\"1/2\\" missed quot(\\")!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer unclosed html <html title/> 1`] = `
Object {
  "actual": "<html title=\\"title\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
  "warning": Array [
    "[xmldom warning]	attribute \\"title\\" missed value!! \\"title\\" instead!!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer unclosed html <html title= 1/> 1`] = `
Object {
  "actual": "<html title=\\"1\\" xmlns=\\"http://www.w3.org/1999/xhtml\\"></html>",
  "warning": Array [
    "[xmldom warning]	attribute \\"1\\" missed quot(\\")!
@#[line:1,col:1]",
  ],
}
`;

exports[`html normalizer unclosed html <html><meta><link><img><br><hr><input></html> 1`] = `
Object {
  "actual": "<html xmlns=\\"http://www.w3.org/1999/xhtml\\"><meta/><link/><img/><br/><hr/><input/></html>",
}
`;
